File: quintic/kernels/cycloidal.py
import numpy as np
from typing import Tuple, Union, cast


class Cycloidal:
    """
    Implements Cycloidal Motion (VDI 2143: Bestehorn / Sinusoid).
    Excellent for high speed, low vibration.
    A_max = 2*pi (~6.28), J_max = 4*pi^2 (~39.48).

    Formula: s(tau) = tau - (1/2pi)*sin(2pi*tau)
    """

    def __init__(
        self,
        master_start: float,
        master_end: float,
        slave_start: float,
        slave_end: float,
    ):
        self.master_start = master_start
        self.master_end = master_end
        self.slave_start = slave_start
        self.slave_end = slave_end

    def generate_table(self, resolution: int = 100) -> list[dict]:
        if resolution < 2:
            return []
        thetas = np.linspace(self.master_start, self.master_end, resolution)
        s, v, a, j = self.calculate(thetas)  # type: ignore

        # Explicit casting for type checking
        s_arr = cast(np.ndarray, s)
        v_arr = cast(np.ndarray, v)
        a_arr = cast(np.ndarray, a)
        j_arr = cast(np.ndarray, j)

        return [
            {
                "theta": float(t),
                "s": float(si),
                "v": float(vi),
                "a": float(ai),
                "j": float(ji),
            }
            for t, si, vi, ai, ji in zip(thetas, s_arr, v_arr, a_arr, j_arr)
        ]

    def calculate(
        self, theta: Union[float, np.ndarray]
    ) -> Union[
        Tuple[float, float, float, float],
        Tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray],
    ]:
        theta_arr = np.atleast_1d(theta)
        beta = self.master_end - self.master_start
        if beta == 0:
            zeros = np.zeros_like(theta_arr, dtype=np.float64)
            return zeros, zeros, zeros, zeros

        tau = (theta_arr - self.master_start) / beta
        h = self.slave_end - self.slave_start

        # Valid range mask
        mask_valid = (tau >= 0) & (tau <= 1)

        s = np.zeros_like(theta_arr, dtype=np.float64)
        v = np.zeros_like(theta_arr, dtype=np.float64)
        a = np.zeros_like(theta_arr, dtype=np.float64)
        j = np.zeros_like(theta_arr, dtype=np.float64)

        # Boundary filling
        s[tau > 1] = self.slave_end
        s[tau < 0] = self.slave_start

        if np.any(mask_valid):
            t = tau[mask_valid]
            TwoPi = 2 * np.pi

            # Displacement: s = h * (tau - 1/2pi * sin(2pi*tau))
            s[mask_valid] = self.slave_start + h * (t - (1 / TwoPi) * np.sin(TwoPi * t))

            # Velocity: v = (h/beta) * (1 - cos(2pi*tau))
            v[mask_valid] = (h / beta) * (1 - np.cos(TwoPi * t))

            # Acceleration: a = (h/beta^2) * 2pi * sin(2pi*tau)
            a[mask_valid] = (h / (beta**2)) * TwoPi * np.sin(TwoPi * t)

            # Jerk: j = (h/beta^3) * 4pi^2 * cos(2pi*tau)
            j[mask_valid] = (h / (beta**3)) * (TwoPi**2) * np.cos(TwoPi * t)

        if np.isscalar(theta):
            return float(s[0]), float(v[0]), float(a[0]), float(j[0])
        return s, v, a, j

File: Quintic.Wpf/Core/Kernels/Cycloidal.cs
using Quintic.Wpf.Core.Models;
using System;
using System.Collections.Generic;

namespace Quintic.Wpf.Core.Kernels
{
    public class Cycloidal : IMotionKernel
    {
        private readonly double _masterStart;
        private readonly double _masterEnd;
        private readonly double _slaveStart;
        private readonly double _slaveEnd;

        public Cycloidal(double masterStart, double masterEnd, double slaveStart, double slaveEnd)
        {
            _masterStart = masterStart;
            _masterEnd = masterEnd;
            _slaveStart = slaveStart;
            _slaveEnd = slaveEnd;
        }

        public CamPoint Calculate(double theta)
        {
            double beta = _masterEnd - _masterStart;
            if (Math.Abs(beta) < 1e-9) return new CamPoint(theta, _slaveStart, 0, 0, 0);

            // Normalize theta to tau (0..1)
            double tau = (theta - _masterStart) / beta;

            // Handle out of bounds
            if (tau < 0) tau = 0;
            if (tau > 1) tau = 1;

            double h = _slaveEnd - _slaveStart;
            double TwoPi = 2 * Math.PI;

            // Displacement: s = h * (tau - 1/2pi * sin(2pi*tau))
            double s_norm = tau - (1.0 / TwoPi) * Math.Sin(TwoPi * tau);
            double s = _slaveStart + h * s_norm;

            // Velocity: v = (h/beta) * (1 - cos(2pi*tau))
            double v_norm = 1.0 - Math.Cos(TwoPi * tau);
            double v = (h / beta) * v_norm;

            // Acceleration: a = (h/beta^2) * 2pi * sin(2pi*tau)
            double a_norm = TwoPi * Math.Sin(TwoPi * tau);
            double a = (h / (beta * beta)) * a_norm;

            // Jerk: j = (h/beta^3) * 4pi^2 * cos(2pi*tau)
            double j_norm = (TwoPi * TwoPi) * Math.Cos(TwoPi * tau);
            double j = (h / (beta * beta * beta)) * j_norm;

            return new CamPoint(theta, s, v, a, j);
        }

        public List<CamPoint> GenerateTable(int resolution)
        {
            var points = new List<CamPoint>(resolution);
            if (resolution < 2) return points;

            double step = (_masterEnd - _masterStart) / (resolution - 1);
            
            for (int i = 0; i < resolution; i++)
            {
                double theta = _masterStart + i * step;
                if (i == resolution - 1) theta = _masterEnd;
                points.Add(Calculate(theta));
            }

            return points;
        }
    }
}
